File: backend/app/schemas/technician_performance.py
from datetime import date

from pydantic import BaseModel

from app.schemas.dashboard import ChartPoint


class TechnicianPerformanceSummary(BaseModel):
    technician_id: int
    full_name: str
    role: str

    # Technician-only fields — populated for technician rows, left at their
    # defaults (None / 0) for chef/admin rows. Unchanged from the
    # technician-only version of this schema.
    total_interventions: int = 0
    completed_interventions: int = 0
    pending_interventions: int = 0
    rejected_interventions: int = 0
    warranty_interventions: int = 0
    total_points: int = 0
    average_duration_minutes: float = 0.0
    planned_count: int = 0
    completed_vs_planned_ratio: float = 0.0
    colleague_participation_count: int = 0
    next_planned_date: date | None = None
    next_planned_client_name: str | None = None

    # Chef/admin-only fields — populated for chef (technical approvals) and
    # admin (administrative approvals) rows, left None for technician rows.
    approvals_processed: int | None = None
    approvals_rejected: int | None = None
    avg_turnaround_minutes: float | None = None


class TechnicianPerformanceDetail(TechnicianPerformanceSummary):
    first_name: str
    last_name: str
    email: str
    phone: str | None
    active: bool
    monthly_activity_chart: list[ChartPoint]
    weekly_activity_chart: list[ChartPoint]

File: frontend/src/types/technicianPerformance.ts
import type { ChartPoint } from "./dashboard";

export interface TechnicianPerformanceSummary {
  technician_id: number;
  full_name: string;
  role: string;

  // Technician-only fields — populated for technician rows, left at their
  // defaults (0) for chef/admin rows.
  total_interventions: number;
  completed_interventions: number;
  pending_interventions: number;
  rejected_interventions: number;
  warranty_interventions: number;
  total_points: number;
  average_duration_minutes: number;
  planned_count: number;
  completed_vs_planned_ratio: number;
  colleague_participation_count: number;
  next_planned_date: string | null;
  next_planned_client_name: string | null;

  // Chef/admin-only fields — populated for chef (technical approvals) and
  // admin (administrative approvals) rows, null for technician rows.
  approvals_processed: number | null;
  approvals_rejected: number | null;
  avg_turnaround_minutes: number | null;
}

export interface TechnicianPerformanceDetail extends TechnicianPerformanceSummary {
  first_name: string;
  last_name: string;
  email: string;
  phone: string | null;
  active: boolean;
  monthly_activity_chart: ChartPoint[];
  weekly_activity_chart: ChartPoint[];
}
